File: src_backend/src/main/java/com/tieuluan/backend/service/DictionaryService.java
package com.tieuluan.backend.service;

import com.tieuluan.backend.model.Dictionary;
import com.tieuluan.backend.repository.DictionaryRepository;
import lombok.Data;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

/**
 * Service tra cứu từ điển offline
 * Thay thế việc gọi Gemini AI cho định nghĩa từ
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class DictionaryService {

    private final DictionaryRepository dictionaryRepository;

    /**
     * Tra cứu từ vựng - trả về đầy đủ thông tin
     */
    public DictionaryLookupResult lookup(String word) {
        log.info("📖 Looking up word: '{}'", word);

        if (word == null || word.trim().isEmpty()) {
            return DictionaryLookupResult.notFound(word, "Từ không được để trống");
        }

        String cleanWord = word.trim().toLowerCase();

        Optional<Dictionary> dictOpt = dictionaryRepository.findByWordIgnoreCase(cleanWord);

        if (dictOpt.isPresent()) {
            Dictionary dict = dictOpt.get();
            log.info("✅ Found word '{}' in dictionary", cleanWord);

            return DictionaryLookupResult.builder()
                    .found(true)
                    .word(dict.getWord())
                    .partOfSpeech(dict.getPartOfSpeech())
                    .partOfSpeechVi(dict.getPartOfSpeechVi())
                    .phonetic(dict.getPhonetic())
                    .definitions(dict.getDefinitions())
                    .meanings(dict.getMeanings())
                    .source(dict.getSource())
                    .build();
        } else {
            log.warn("⚠️ Word '{}' not found in dictionary", cleanWord);
            return DictionaryLookupResult.notFound(cleanWord, "Từ không có trong từ điển");
        }
    }

    /**
     * Gợi ý từ khi người dùng đang gõ (autocomplete)
     */
    public List<String> suggest(String prefix) {
        if (prefix == null || prefix.trim().length() < 2) {
            return List.of();
        }

        String cleanPrefix = prefix.trim().toLowerCase();
        log.info("🔍 Suggesting words starting with: '{}'", cleanPrefix);

        List<Dictionary> results = dictionaryRepository.findByWordStartingWith(cleanPrefix);

        return results.stream()
                .map(Dictionary::getWord)
                .collect(Collectors.toList());
    }

    /**
     * Tìm kiếm từ chứa keyword
     */
    public List<DictionaryLookupResult> search(String keyword) {
        if (keyword == null || keyword.trim().length() < 2) {
            return List.of();
        }

        String cleanKeyword = keyword.trim().toLowerCase();
        log.info("🔎 Searching words containing: '{}'", cleanKeyword);

        List<Dictionary> results = dictionaryRepository.findByWordContaining(cleanKeyword);

        return results.stream()
                .map(dict -> DictionaryLookupResult.builder()
                        .found(true)
                        .word(dict.getWord())
                        .partOfSpeech(dict.getPartOfSpeech())
                        .partOfSpeechVi(dict.getPartOfSpeechVi())
                        .phonetic(dict.getPhonetic())
                        .definitions(dict.getDefinitions())
                        .meanings(dict.getMeanings())
                        .source(dict.getSource())
                        .build())
                .collect(Collectors.toList());
    }

    /**
     * Kiểm tra từ có tồn tại trong từ điển không
     */
    public boolean exists(String word) {
        if (word == null || word.trim().isEmpty()) {
            return false;
        }
        return dictionaryRepository.existsByWord(word.trim());
    }

    /**
     * Lấy thống kê từ điển
     */
    public DictionaryStats getStats() {
        long totalWords = dictionaryRepository.countAll();
        return new DictionaryStats(totalWords);
    }

    // ================== DTOs ==================

    @Data
    @lombok.Builder
    public static class DictionaryLookupResult {
        private boolean found;
        private String word;
        private String partOfSpeech;      // noun, verb, adjective...
        private String partOfSpeechVi;    // danh từ, động từ...
        private String phonetic;          // /ˈbæŋk/
        private String definitions;       // English definition
        private String meanings;          // Vietnamese meaning
        private String source;            // vi+en, en, vi
        private String errorMessage;

        public static DictionaryLookupResult notFound(String word, String message) {
            return DictionaryLookupResult.builder()
                    .found(false)
                    .word(word)
                    .errorMessage(message)
                    .build();
        }
    }

    @Data
    @lombok.AllArgsConstructor
    public static class DictionaryStats {
        private long totalWords;
    }
}

File: src_app/lib/services/dictionary_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../config/api_config.dart';

/// Service tra cứu từ điển offline
class DictionaryService {
  /// Tra cứu từ vựng
  static Future<DictionaryLookupResult> lookup(String word) async {
    try {
      final uri = Uri.parse('${ApiConfig.baseUrl}/api/dictionary/lookup')
          .replace(queryParameters: {'word': word});

      print('📖 Looking up: $word');

      final response = await http.get(uri, headers: {
        'Content-Type': 'application/json',
        'ngrok-skip-browser-warning': 'true',
      });

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return DictionaryLookupResult.fromJson(data);
      } else {
        throw Exception('Failed to lookup word: ${response.statusCode}');
      }
    } catch (e) {
      print('❌ Dictionary lookup error: $e');
      rethrow;
    }
  }

  /// Gợi ý từ khi đang gõ (autocomplete)
  static Future<List<String>> suggest(String prefix) async {
    try {
      if (prefix.length < 2) return [];

      final uri = Uri.parse('${ApiConfig.baseUrl}/api/dictionary/suggest')
          .replace(queryParameters: {'prefix': prefix});

      final response = await http.get(uri, headers: {
        'Content-Type': 'application/json',
        'ngrok-skip-browser-warning': 'true',
      });

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.cast<String>();
      } else {
        return [];
      }
    } catch (e) {
      print('❌ Dictionary suggest error: $e');
      return [];
    }
  }

  /// Tìm kiếm từ chứa keyword
  static Future<List<DictionaryLookupResult>> search(String keyword) async {
    try {
      final uri = Uri.parse('${ApiConfig.baseUrl}/api/dictionary/search')
          .replace(queryParameters: {'keyword': keyword});

      final response = await http.get(uri, headers: {
        'Content-Type': 'application/json',
        'ngrok-skip-browser-warning': 'true',
      });

      if (response.statusCode == 200) {
        final List<dynamic> data = jsonDecode(response.body);
        return data.map((e) => DictionaryLookupResult.fromJson(e)).toList();
      } else {
        return [];
      }
    } catch (e) {
      print('❌ Dictionary search error: $e');
      return [];
    }
  }

  /// Kiểm tra từ có tồn tại không
  static Future<bool> exists(String word) async {
    try {
      final uri = Uri.parse('${ApiConfig.baseUrl}/api/dictionary/exists')
          .replace(queryParameters: {'word': word});

      final response = await http.get(uri, headers: {
        'ngrok-skip-browser-warning': 'true',
      });

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return data['exists'] ?? false;
      } else {
        return false;
      }
    } catch (e) {
      return false;
    }
  }

  /// Lấy thống kê từ điển
  static Future<DictionaryStats> getStats() async {
    try {
      final uri = Uri.parse('${ApiConfig.baseUrl}/api/dictionary/stats');

      final response = await http.get(uri, headers: {
        'ngrok-skip-browser-warning': 'true',
      });

      if (response.statusCode == 200) {
        final data = jsonDecode(response.body);
        return DictionaryStats.fromJson(data);
      } else {
        throw Exception('Failed to get stats');
      }
    } catch (e) {
      print('❌ Dictionary stats error: $e');
      rethrow;
    }
  }
}

/// Kết quả tra từ điển
class DictionaryLookupResult {
  final bool found;
  final String? word;
  final String? partOfSpeech;
  final String? partOfSpeechVi;
  final String? phonetic;
  final String? definitions;
  final String? meanings;
  final String? source;
  final String? errorMessage;

  DictionaryLookupResult({
    required this.found,
    this.word,
    this.partOfSpeech,
    this.partOfSpeechVi,
    this.phonetic,
    this.definitions,
    this.meanings,
    this.source,
    this.errorMessage,
  });

  factory DictionaryLookupResult.fromJson(Map<String, dynamic> json) {
    return DictionaryLookupResult(
      found: json['found'] ?? false,
      word: json['word'],
      partOfSpeech: json['partOfSpeech'],
      partOfSpeechVi: json['partOfSpeechVi'],
      phonetic: json['phonetic'],
      definitions: json['definitions'],
      meanings: json['meanings'],
      source: json['source'],
      errorMessage: json['errorMessage'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'found': found,
      'word': word,
      'partOfSpeech': partOfSpeech,
      'partOfSpeechVi': partOfSpeechVi,
      'phonetic': phonetic,
      'definitions': definitions,
      'meanings': meanings,
      'source': source,
      'errorMessage': errorMessage,
    };
  }
}

/// Thống kê từ điển
class DictionaryStats {
  final int totalWords;

  DictionaryStats({required this.totalWords});

  factory DictionaryStats.fromJson(Map<String, dynamic> json) {
    return DictionaryStats(
      totalWords: json['totalWords'] ?? 0,
    );
  }
}